staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go

```diff
@@ -43,11 +43,11 @@ func WithFailedAuthenticationAudit(failedHandler http.Handler, sink audit.Sink,
 			return
 		}
 
-		if ac == nil || ac.Event == nil {
+		if !ac.Enabled() {
```

Comment:
It seems the `Enabled` function is used outside of its package where the caller does not not have the pre-evaluation constraints we have inside the `audit` package, so for future maintainability i think the following might be useful:
```
func (ac *AuditContext) Enabled() bool {
	return ac != nil && len(ac.RequestAuditConfig.Level) > 0 && ac.RequestAuditConfig.Level != auditinternal.LevelNone
}
```

inside the `audit` package, where we have to be mindful of the pre-evaluation constraint, we can use the following private function:
```
func (ac *AuditContext) shouldAddAnnotation() bool {
	switch {
	case ac == nil:
		return false
	case ac.RequestAuditConfig.Level == "":
		return true
	}
	return ac.Enabled()
}
```

```
func AddAuditAnnotation(ctx context.Context, key, value string) {
	ac := AuditContextFrom(ctx)
	if !ac.shouldAddAnnotation() {
		return
	}
...
```

- this approach does not leak the constraint outside of the `audit` package, the invoker of `Enabled` is aware that policy has been evaluated
- `AddAnnotation*` functions have the pre-evaulation constraint and these functions can resort to an internal `shouldAddAnnotation`
Thoughts?

Reply:
I'm afraid I don't entirely follow this comment.

In the next audit refactor pass, I'd like to make `ac.Event` non-exported, and enforce that all event handling be done within the audit package, which I think will make this comment moot.

If you're OK with it, I propose we leave this as-is for now (the logic matches the previous logic in practice, as the event initialization & policy evaluation happen in the same step), and revisit this in the follow-up PR. WDYT?

Reply:
yes, that sounds good to me.